Getting-Started/Code/Debugging/Logging/index-v8.md

```diff
@@ -0,0 +1,250 @@
+}
+```
+
+To learn more about structured logging and message templates you can read more about it over on the https://messagetemplates.org website or alternatively watch this video from Serilog creator - https://www.youtube.com/watch?v=OhmNp8UPEEg
+
+## Writing to the log
+Umbraco writes log messages, but you are also able to use the Umbraco logger to write the log file as needed, so you can get further insights and details about your implementation.
+
+Here is a simple example of using the logger to write an Information message to the log which will contain one property of **Name** which will output the name variable that is passed into the method
+
+```csharp
+using Umbraco.Web.WebApi;
+using Umbraco.Core.Logging;
+
+namespace MyNamespace
+{
+    public class MyApiController : UmbracoApiController
+    {
+        public string GetSayHello(string name)
+        {
+            Logger.Info<MyApiController>("We are saying hello to {Name}", name);
+            return $"Hello {name}";
+        }
+    }
+}
+```
+
+The incorrect way to log the message would be use string interpolation or string concatanation such as
+
+```csharp
+//Do not use :(
+Logger.Info(typeof(MyApiController), $"We are saying hello to {name}");
+Logger.Info<MyApiController>($"We are saying hello to {name}");
+
+//Do not use :(
+Logger.Info(typeof(MyApiController), "We are saying hello to " + name);
+Logger.Info<MyApiController>("We are saying hello to " + name);
+```
+
+The above examples will write to the log file, however we will not get a seperate property logged with the message and we have no easy way to search for all log messages of this type.
+
+Where as the previous example we would be able to find all log messages that use the message template `We are saying hello to {Name}`
+
+If you are writing custom code and wish to use the logger, it's available in the following places:
+* SurfaceController
+* UmbracoApiController
+* UmbracoAuthorizedApiController
+
+If you wish to use it in any other place you can retrieve it by using the `Current.Logger` from the `Umbraco.Core.Composing` namespace.
+
+
+## Log Levels
+Serilog uses levels as the primary means for assigning importance to log events. The levels in increasing order of importance are:
+
+1. **Verbose** - tracing information and debugging minutiae; generally only switched on in unusual situations
+1. **Debug** - internal control flow and diagnostic state dumps to facilitate pinpointing of recognised problems
+1. **Information** - events of interest or that have relevance to outside observers; the default enabled minimum logging level
+1. **Warning** - indicators of possible issues or service/functionality degradation
+1. **Error** - indicating a failure within the application or connected system
+1. **Fatal** - critical errors causing complete failure of the application
+
+The default log levels we ship with in Umbraco v8.0+ are:
+* .txt file **Debug**
+* .json file **Verbose**
+
```

Comment:
Do we want to put anything in here about over-logging? or is serilog super performant! Log4Net can cause havoc if it's set to Debug, and there is an abundance of debug messages logged...

Reply:
I am not sure if or how much more performant Serilog is versus Log4Net and yet to have a site where I have heavily stressed tested this yet @marcemarc in the wild whilst V8 still in active dev.

What would happen with Log4Net when hammered with logs, does it cause the site to stop rendering or slow down requests or just kill your disk IO?!

We can write something about not over logging & I can change the defaults we ship in V8 to be higher if that is more generally prefered, but it would be good for me to test/verify somehow if this is more performant than Log4Net or not, but I would need to know what issues to look for/compare to.

Reply:
@warrenbuckley yes, the site requests slow down, presuming because of disk IO, generally, now in production we run a site with logging levels set to ERROR, (small logs, see errors easily) -  then switch to DEBUG if there is an issue to debug, that way you can 'over log' in DEBUG, eg the more logging the better!, but without needlessly logging info when all is running fine. but it's a trade-off as I guess having that debug info there for people new to Umbraco from the get-go, means it's discoverable as a site is developed, ....so maybe it's less of a Tip for the documentation of serilog, and more relevant for a page on 'hardening umbraco for production' ?

Reply:
Yes perhaps it's a tip or health check instead?!
Some other Serilog sinks batch the logs and send them off to the destination, so another hardening tip could be to store logs in a different source than on disk.

What are your thoughts on the default log levels that are shipped in V8 for the txt & json file @marcemarc ?!

Reply:
@jmayntzhusen I agree with you but the plan is to have an inbuilt logviewer built into the core of Umbraco so you can easily search & filter these log messages. It's currently a WIP and will be due for a tech review soon by the core D-Team

Here is an early sneak peek
![image](https://user-images.githubusercontent.com/1389894/48411236-ddca2f00-e738-11e8-863a-dbab7d77f0eb.png)

![image](https://user-images.githubusercontent.com/1389894/48411331-2a156f00-e739-11e8-80f8-d80f4ab313e5.png)